### ISSUE TITLE
Hack syntax highlighting broken by callable types (at least in shapes)

### ISSUE THREAD

**Issue by fredemmott** [facebook/nuclide#765]
### Issue and Steps to Reproduce

Open https://github.com/fredemmott/hack-router-codegen/blob/v1.0.0rc2/src/Codegen.php in nuclide

Syntax highlighting is broken from `(function` onwards; happening in several other files too - all are inside a shape in a class type constant.

![screen shot 2016-10-05 at 6 14 36 pm](https://cloud.githubusercontent.com/assets/360927/19137339/e79342e8-8b27-11e6-834f-ba971c4a4bb0.png)
### Expected Behavior

Correct syntax highlighting, or worst case, none for the declaration but fixed at the end of it.
### Actual Behavior

grey text for quite some time after
### Versions
- Atom: 1.10.2
- Nuclide: 0.174.0 (not a regression, had same with 0.171 at least)
- Client OS: macOS Sierra
- Server OS (optional):
### Additional Details
- Installed packages (`apm ls --installed`):

```
Community Packages (14) /Users/fred/.atom/packages
├── ex-mode@0.13.0
├── git-grep-provider@0.9.3
├── haskell-grammar@0.4.0
├── highlight-selected@0.11.2
├── language-babel@2.43.0
├── language-ini@1.16.0
├── language-ocaml@1.1.2
├── language-swift@0.4.0
├── language-thrift@1.0.2
├── merge-conflicts@1.4.4
├── nuclide@0.174.0
├── sort-lines@0.14.0
├── tool-bar@1.0.1
└── vim-mode@0.65.1

└── (empty)
```

**Comment by jesseschalken** [comment]
Also broken when `function` occurs in a type parameter to a generic interface method:

![image](https://cloud.githubusercontent.com/assets/1559204/19137592/c0ea3518-8bc0-11e6-994e-518575c4d2df.png)

Removing the type parameter on the method fixes it. For some reason generic methods/functions get different colors for their name than non-generic ones, which is possibly related.

I defined these types to work around it (the problem also occurs in the definition of these types):

```
type fun0<+T> = (function(): T);
type fun1<-T1, +T> = (function(T1): T);
type fun2<-T1, -T2, +T> = (function(T1, T2): T);
type fun3<-T1, -T2, -T3, +T> = (function(T1, T2, T3): T);
```

It was bugs like this that [led me to consider](https://www.facebook.com/groups/hhvm.general/permalink/683219185164811/?comment_id=684665041686892) using `token_get_all()` to create a syntax highlighter in PHP/Hack to test Nuclide's highlighter against.

**Comment by shushz** [comment]
Yeah, from what I've been told Hack is not a regular language, and regex-based syntax highlighting will never be completely correct. There are plans to move it to a dynamic grammar, but it's not something to expect soon as there are significant gaps to bridge.

**Comment by jesseschalken** [comment]
I can't think of any programming languages that are regular languages, since a regular language wouldn't be able to have nested parenthesis.

[Atom's language grammars](https://manual.macromates.com/en/language_grammars) (based on TextMate) allow recursive rules which means it can parse non-regular languages. Vim's syntax definitions are similar. It'd be pretty useless otherwise.

**Comment by jesseschalken** [comment]
The brokenness I mentioned above with function parameters was recently fixed (probably by https://github.com/facebook/nuclide/commit/2c7e6aedbaadce16e1f2950224370d8a565113e3 ?) but the original issue with `(function` in a shape still stands.

**Comment by fredemmott** [comment]
Still broken in nuclide@0.200.0

**Comment by fredemmott** [comment]
This was fixed by ccec96c1c629aa8931df74f2929aa9faaeaa383a
